scripts/trader.py

```diff
@@ -12,46 +13,79 @@ class Trade:
     quantity: int
 
 # Placeholder function to query Polygon API for stock price
-def get_stock_price(ticker: str, date: str) -> float:
+def get_stock_price(ticker: str, date: datetime) -> float:
     """Fetches stock price from Polygon API for a given ticker and date."""
     # TODO: Implement API call
-    return 100.0  # Placeholder price
+    stock = yf.Ticker(ticker)
+    start = date.strftime('%Y-%m-%d')
+    end = (date + timedelta(days=1)).strftime('%Y-%m-%d')
+    data = stock.history(start=start, end=end)
+    if not data.empty:
+        return round(float(data['Close'].iloc[0]), 2)
+    else:
+        raise ValueError("No data found for the given date")
 
 # Function to execute a trade and update portfolio
-def execute_trade(portfolio: Dict[str, Dict[str, float]], trade: Trade):
+def execute_trade(portfolio: Dict[str, Dict[str, Any]], trade: Trade, balance: float) -> Tuple[Dict[str, Dict[str, Any]], float]:
     """Updates portfolio based on the trade action."""
 
-    # TODO: call the "get_stock_price" function to get the stock price for the trade date
-    
-    # TODO: check if the trade ticker is in the portfolio, if not add it
+    # call the "get_stock_price" function to get the stock price for the trade date
+    price = get_stock_price(trade.ticker, trade.date)
+
+    # check if the trade ticker is in the portfolio, if not add it
     #   initialize the stock position with average cost, shares, and trades list
+    if trade.ticker not in portfolio:
+        portfolio[trade.ticker] = {
+            "average_cost": 0.0,
+            "shares": 0,
+            "trades": []
+        }
     
-    # TODO: if the action is 'buy', add number of shares; if 'sell', remove shares
+    # if the action is 'buy', add number of shares; if 'sell', remove shares
     #   update the average cost based on the total cost of shares
+    if trade.action == 'buy':
+        total_cost = price * trade.quantity
+        total_value = portfolio[trade.ticker]["average_cost"] * portfolio[trade.ticker]["shares"] + total_cost
```

Comment:
This line of code is pretty unreadable... We're calling `portfolio[trade.ticker]`, a bunch of times. Why not just assign a variable to this, get the values we need, reassign `portfolio[trade.ticker]` to the temp variable? 

```suggestion
        current_position = portfolio[trade.ticker]
        total_value = current_position["average_cost"] * current_position["shares"] + total_cost
```

and at the end
```suggestion
        portfolio[trade.ticker] = current_holdings
```

Reply:
Don't accept this suggestion as it. You'll have to pull `current_position = portfolio[trade.ticker]` to before the if statement for scope, replace all references of portfolio to `current_position`, then assign `portfolio[trade.ticker] = current_position` back at the end, outside of your if statement

Reply:
Fixed the issue. I comitted and pushed the changes.